core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -292,16 +287,23 @@ public void registerSuggester(String key, Suggester<?> suggester) {
     }
 
     /**
-     * Register a new ScoreFunctionParser.
+     * Register a new ScoreFunctionBuilder. Registration does two things:
+     * <ul>
+     * <li>Register the {@link ScoreFunctionParser} which parses XContent into a {@link ScoreFunctionBuilder} using its {@link ParseField}
+     * </li>
+     * <li>Register the {@link Writeable.Reader} which reads a stream representation of the builder under a single name. This name must be
+     * the same name as the builder returns for getWriteableName.</li>
+     * </ul>
      */
-    public <T extends ScoreFunctionBuilder<T>> void registerScoreFunction(ScoreFunctionParser<T> parser, Writeable.Reader<T> reader) {
-        for (String name: parser.getNames()) {
+    public <T extends ScoreFunctionBuilder<T>> void registerScoreFunction(ParseField xContentField, ScoreFunctionParser<T> parser,
+            String writeableName, Writeable.Reader<T> reader) {
+        for (String name: xContentField.getAllNamesIncludedDeprecated()) {
```

Comment:
Right. Should I do that now? I was thinking maybe another PR because this one is already huge and getting the ParseField in the right place gets us most of the way there. I'll see how much it is.